File: 2022/06/cpp/06.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long          ll;
typedef vector<int>        vi;
typedef vector<bool>       vb;
typedef pair<int, int>     ii;

const int INF = 0x3f3f3f3f;
const ll  INFLL = 0x3f3f3f3f3f3f3f3fll;

#define each(x, s)  for(auto& x : s)
#define loop(x)     for(int i = 0;i < x;i++)
#define vloop(v, x) for(int v = 0;v < x;v++)
#define iter(a)     a.begin(), a.end()
#define riter(a)    a.rbegin(), a.rend()
#define endl        "\n"
#define avg(l, r)   (l + r) / 2

int find(string& s, int cnt) {
  vi last(26, -1);
  int start = 0;
  for (int i = 0;s[i];i++) {
    int c = s[i] - 'a';
    start = max(start, last[c] + 1);
    if (i - start == cnt - 1) return i + 1;
    last[c] = i;
  }
  return 0;
}

int task1(string& s) {
  return find(s, 4);
}

int task2(string& s) {
  return find(s, 14);
}

int main() {
  ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

  string s;
  cin >> s;
  cout << "Task 1: " << task1(s) << endl;
  cout << "Task 2: " << task2(s) << endl;
  return 0;
}

File: 2022/06/haskell/06.hs
{- stack script --resolver lts-14.20
 --package containers -}
module Main where

import qualified Data.Map as M
import Data.Maybe

find :: Int -> String -> Int
find x s = find' M.empty x 1 1 s
  where 
    find' :: M.Map Char Int -> Int -> Int -> Int -> String -> Int
    find' last cnt start i (c:cs) | i - start' + 1 == cnt = i
                                  | otherwise             = find' last' cnt start' (i + 1) cs
      where
        start' = max start $ (+1) . fromMaybe (-1) $ M.lookup c last
        last'  = M.insert c i last

task1 :: String -> Int
task1 = find 4

task2 :: String -> Int
task2 = find 14

main :: IO ()
main = do
  input <- getContents
  putStr "Task 1: "
  print $ task1 input
  putStr "Task 2: "
  print $ task2 input

